### PR TITLE
add video capture parameters 

### PR DESCRIPTION
**Pull request by MaximMilashchenko**
relates #16766

@alalek , @mshabunin , @allnes

```
Xforce_builders_only=linux,mac,docs
```

### CODE REVIEW

File: modules/videoio/src/cap_aravis.cpp [hunk 1]
```diff
@@ -102,7 +102,7 @@ class CvCaptureCAM_Aravis : public CvCapture
         close();
     }
 
-    virtual bool open(int);
+    virtual bool open(int, const cv::VideoCaptureParameters&);
```
Comment (alalek) [inline modules/videoio/src/cap_aravis.cpp:105]: Lets avoid modification of all backends implementation.
Use virtual function of base class instead.

Modify 1-2 backend which are interested and can be used as sample.

**Comment by alalek** [comment]
> Merge branch 'master'

Avoid such merge commits in PRs. Prefer to keep 1 commit with proper author/e-mail information instead.

Use `git rebase -i` to maintain that.

File: modules/videoio/src/cap_aravis.cpp (hunk 1 above)
Comment (MaximMilashchenko) [reply modules/videoio/src/cap_aravis.cpp:105]: add the open(..., const cv :: VideoCaptureParameters &) function to the base class and implement it only in the required backends (example: gstreamer, msmf)?

Comment (alalek) [reply modules/videoio/src/cap_aravis.cpp:105]: yes.

Also check similar patch for `VideoWriter` from here: #16766

Comment (MaximMilashchenko) [reply modules/videoio/src/cap_aravis.cpp:105]: Still don't get it. Do you mean that you need to take all the open () to the top in the base class? Can you tell us more about your idea. When implementing the parameters of the video writer, the implementation of the open () function and the extraction of parameters remains at the backend level.

Comment (alalek) [reply modules/videoio/src/cap_aravis.cpp:105]: Add common base class method:

```
bool CvCapture::open(int idx, const cv::VideoCaptureParameters& params)
{
    bool res = open(idx);
    if (res && !params.empty())
    {
        res &= setParameters(params);
        if (!res)
            close();
    }
    return res;
}
```

---

Derived classes which doesn't care about new API doesn't redefine this method.

Only FFmpeg / GStreamer / MSMF backends may have such method as an example (but do not bring large functionality there - it is not in scope of this PR). If you don't understand then don't add this overload except base classes.

**Review by alalek — COMMENTED** [review]
Current patch has these common mistakes:
- no tests (especially for new features)
- massive code changes (which should be avoided): 28 changed files
- lack of logic for handling of error cases (like silent ignoring of parameters)

File: modules/videoio/src/cap_android_mediandk.cpp [hunk 2]
```diff
@@ -240,7 +240,7 @@ class AndroidMediaNdkCapture : public IVideoCapture
 
 /****************** Implementation of interface functions ********************/
 
-Ptr<IVideoCapture> cv::createAndroidCapture_file(const std::string &filename) {
+Ptr<IVideoCapture> cv::createAndroidCapture_file(const std::string &filename, const cv::VideoCaptureParameters&) {
```
[the remark this answers is not in the document]
Comment (MaximMilashchenko) [reply modules/videoio/src/cap_android_mediandk.cpp]: I ignore parameters in backends that don't have a open() function, or where the base class is CVCapture. Should I add a message about the lack of support for params in this backend? I can also add to the CVCapture the same overload of open functions as in IVideoCapture.

[the remark this answers is not in the document]
Comment (alalek) [reply modules/videoio/src/cap_android_mediandk.cpp]: These changes are dropped.
Fetch latest changes from GitHub with updated handling of such cases.

**Review by alalek — APPROVED** [review]

**Pull request merged**